lib/AST/LocalizationFormat.cpp

```diff
@@ -143,31 +132,46 @@ void YAMLLocalizationProducer::forEachAvailable(
   }
 }
 
+llvm::Optional<uint32_t> readID(llvm::yaml::IO &io) {
+  LocalDiagID diagID;
+  io.mapRequired("id", diagID);
+  if (diagID == LocalDiagID::NumDiags)
+    return llvm::None;
+  return static_cast<uint32_t>(diagID);
+}
+
 template <typename T, typename Context>
 typename std::enable_if<llvm::yaml::has_SequenceTraits<T>::value, void>::type
-readYAML(llvm::yaml::IO &io, T &Seq, bool, Context &Ctx) {
+readYAML(llvm::yaml::IO &io, T &Seq, T &unknownIDs, bool, Context &Ctx) {
   unsigned count = io.beginSequence();
-  if (count)
+  if (count) {
     Seq.resize(LocalDiagID::NumDiags);
+  }
+
+  std::string currentMsg;
```

Comment:
@xedin I think there's no need for the `DiagnosticNode` anymore since I've a function that will read the `id` for me and in L157 https://github.com/apple/swift/pull/33337/files#diff-594606b3ff2e7321e3fac5cc36c08dcbR157 I'll be manually reading the diagnostic message as a string. WDYT?

Reply:
Sure, that makes sense, if we are not doing that anymore let's not declare `currentMsg` in outer scope like that. I believe we could do it this way `io.mapRequired(Seq[*id]);` and store a string directly into diagnostic vector instead of copying it from one place to another so the code is going to look like this:

```c++
io.beginMapping();

auto id = LocalizationInput::readID(io);
if (!id) {
  std::string unknownID, message;

  // Read "raw" id since it doesn't exist in `.def` file.
  io.mapRequired("id", unknownID);
  io.mapRequired("msg", message);

  unknownIDs.push_back(unknownID);
} else {
  io.mapRequired("msg", Seq[*id]);
}

io.endMapping();
io.postflightElement(SaveInfo);
```